x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/SetIndexMetadataPropertyAction.java

```diff
@@ -0,0 +1,197 @@
+        ) {
+            super(
+                SetIndexMetadataPropertyAction.NAME,
+                transportService,
+                clusterService,
+                threadPool,
+                actionFilters,
+                SetIndexMetadataPropertyRequest::new,
+                indexNameExpressionResolver,
+                SetIndexMetadataPropertyResponse::new,
+                threadPool.executor(ThreadPool.Names.MANAGEMENT)
+            );
+            this.setIndexMetadataPropertyTaskMasterServiceTaskQueue = clusterService.createTaskQueue(
+                "set-index-metadata-property-task-queue",
+                Priority.LOW,
+                SET_INDEX_METADATA_PROPERTY_TASK_VOID_SIMPLE_BATCHED_EXECUTOR
+            );
+        }
+
+        private static final SimpleBatchedExecutor<
+            SetIndexMetadataPropertyAction.TransportAction.SetIndexMetadataPropertyTask,
+            Map<String, String>> SET_INDEX_METADATA_PROPERTY_TASK_VOID_SIMPLE_BATCHED_EXECUTOR = new SimpleBatchedExecutor<>() {
+                @Override
+                public Tuple<ClusterState, Map<String, String>> executeTask(
+                    SetIndexMetadataPropertyAction.TransportAction.SetIndexMetadataPropertyTask task,
+                    ClusterState clusterState
+                ) {
+                    return task.execute(clusterState);
+                }
+
+                @Override
+                public void taskSucceeded(
+                    SetIndexMetadataPropertyAction.TransportAction.SetIndexMetadataPropertyTask task,
+                    Map<String, String> value
+                ) {
+                    task.success(value);
+                }
+            };
+
+        static class SetIndexMetadataPropertyTask implements ClusterStateTaskListener {
+            private final ActionListener<SetIndexMetadataPropertyResponse> listener;
+            private final Index index;
+            private final String key;
+            @Nullable
+            private final Map<String, String> expected;
+            @Nullable
+            private final Map<String, String> value;
+
+            SetIndexMetadataPropertyTask(
+                ActionListener<SetIndexMetadataPropertyResponse> listener,
+                Index index,
+                String key,
+                @Nullable Map<String, String> expected,
+                @Nullable Map<String, String> value
+            ) {
+                this.listener = listener;
+                this.index = index;
+                this.key = key;
+                this.expected = expected;
+                this.value = value;
+            }
+
+            Tuple<ClusterState, Map<String, String>> execute(ClusterState state) {
+                IndexMetadata indexMetadata = state.metadata().getIndexSafe(index);
+                Map<String, String> existingValue = indexMetadata.getCustomData(key);
+                if (Objects.equals(expected, existingValue)) {
```

Comment:
The action works as a compare-and-swap rather than a simple write.
I think this greatly improves the non-happy path handling when calling this action.